commit fd00e1db80efb2a5b114b0cfb0fd996b9a319e22
Author: Isaac Goldthwaite <igoldthwaite@kududyn.com>
Date:   Wed May 21 15:19:24 2025 -0400

    [automated] Set delta state

diff --git a/.aixcc/README.md b/.aixcc/README.md
new file mode 100644
index 0000000..d3ab8c5
--- /dev/null
+++ b/.aixcc/README.md
@@ -0,0 +1,3 @@
+# Integration Test Challenge
+
+This challenge is used to test CRS integration.
diff --git a/.aixcc/challenge.yaml b/.aixcc/challenge.yaml
new file mode 100644
index 0000000..36bc675
--- /dev/null
+++ b/.aixcc/challenge.yaml
@@ -0,0 +1,27 @@
+metadata_spec_version: v1
+
+name: 'Integration Test Delta Scan Challenge'
+
+# fuzz_tooling project info that pairs to this challenge
+fuzz_tooling_project_name: 'integration-test'
+fuzz_tooling_url: "https://github.com/aixcc-finals/oss-fuzz-aixcc.git"
+fuzz_tooling_ref: "challenge-state/integration-test-delta-01"
+
+challenge_type: delta     # full | delta
+
+base_ref: 4a714359c60858e3821bd478dc846de1d04dc977
+delta_ref: challenges/integration-test-delta-01
+
+# A list of the vulnerabilities included in this challenge
+# This should always match the dir names in .aixcc/vulns/*
+vulnerabilities:
+    - vuln_001
+
+sarif:
+    broadcast: false
+    target: null
+    correct: null
+
+# harness fuzz targets available for this challenge
+harnesses:
+    - 'fuzz_vuln'
diff --git a/.aixcc/test.sh b/.aixcc/test.sh
new file mode 100755
index 0000000..702e46b
--- /dev/null
+++ b/.aixcc/test.sh
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+make test
\ No newline at end of file
diff --git a/.aixcc/vulns/vuln_001/README.md b/.aixcc/vulns/vuln_001/README.md
new file mode 100644
index 0000000..745a955
--- /dev/null
+++ b/.aixcc/vulns/vuln_001/README.md
@@ -0,0 +1,3 @@
+# A sample vuln
+
+This is a simple, sample vuln for the integration test repository.
diff --git a/.aixcc/vulns/vuln_001/blobs/blobs.bin b/.aixcc/vulns/vuln_001/blobs/blobs.bin
new file mode 100644
index 0000000..117a877
--- /dev/null
+++ b/.aixcc/vulns/vuln_001/blobs/blobs.bin
@@ -0,0 +1 @@
+name: AAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBBAAAABBBB
\ No newline at end of file
diff --git a/.aixcc/vulns/vuln_001/patches/good-patch.diff b/.aixcc/vulns/vuln_001/patches/good-patch.diff
new file mode 100644
index 0000000..31f29a9
--- /dev/null
+++ b/.aixcc/vulns/vuln_001/patches/good-patch.diff
@@ -0,0 +1,16 @@
+diff --git a/vuln.c b/vuln.c
+index 15206dc..df29443 100644
+--- a/vuln.c
++++ b/vuln.c
+@@ -26,7 +26,8 @@ bool person_info_parse_file(person_info_t * person_info, const char * const in)
+     for (; isspace(in[last_pos]); last_pos++);
+ 
+     // The bug is THE LINE BELOW THIS LINE
+-    strcpy(person_info->name, &in[last_pos]);
++    strncpy(person_info->name, &in[last_pos], MAX_STRLEN-1);
++    person_info->name[MAX_STRLEN-1] = '\0';
+ 
+     return true;
+-}
+\ No newline at end of file
++}
diff --git a/.aixcc/vulns/vuln_001/vuln.yaml b/.aixcc/vulns/vuln_001/vuln.yaml
new file mode 100644
index 0000000..1316655
--- /dev/null
+++ b/.aixcc/vulns/vuln_001/vuln.yaml
@@ -0,0 +1,30 @@
+metadata_spec_version: v1
+
+name: Buffer overflow in person_info_parse_file
+
+author: anonymous
+
+details:
+
+  cwes:
+    - CWE-121
+
+  description: >
+    Failure to check the bounds of a string by using a strcpy
+
+  locations:
+    - path_from_root: 'vuln.c'
+      startLine: 29
+      startColumn: 1
+      endLine: 30
+      endColumn: 1
+
+pov:
+  blob: 'blobs.bin'
+  harness: 'fuzz_vuln'
+  architecture_build_flag: 'x86_64'
+  sanitizer_build_flag: 'address'
+
+patch:
+  good: 'good-patch.diff'
+  bad: 'bad-patch.diff'
diff --git a/vuln.c b/vuln.c
index fb3c49b..15206dc 100644
--- a/vuln.c
+++ b/vuln.c
@@ -25,5 +25,8 @@ bool person_info_parse_file(person_info_t * person_info, const char * const in)
 
     for (; isspace(in[last_pos]); last_pos++);
 
+    // The bug is THE LINE BELOW THIS LINE
+    strcpy(person_info->name, &in[last_pos]);
+
     return true;
 }
\ No newline at end of file
