commit fd00e1db80efb2a5b114b0cfb0fd996b9a319e22
Author: Isaac Goldthwaite <igoldthwaite@kududyn.com>
Date:   Wed May 21 15:19:24 2025 -0400

diff --git a/vuln.c b/vuln.c
index fb3c49b..15206dc 100644
--- a/vuln.c
+++ b/vuln.c
@@ -25,5 +25,8 @@ bool person_info_parse_file(person_info_t * person_info, const char * const in)
 
     for (; isspace(in[last_pos]); last_pos++);
 
+    // The bug is THE LINE BELOW THIS LINE
+    strcpy(person_info->name, &in[last_pos]);
+
     return true;
 }
\ No newline at end of file
